public/pages/workflow_detail/workspace/workspace.tsx

```diff
@@ -3,24 +3,161 @@
 
-export function Workspace() {
-  const { components } = useSelector((state: AppState) => state.workspace);
+// styling
+import 'reactflow/dist/style.css';
+import './reactflow-styles.scss';
+
+// eslint-disable-next-line @typescript-eslint/no-empty-interface
+interface WorkspaceProps {}
+
+export function Workspace(props: WorkspaceProps) {
+  const reactFlowWrapper = useRef(null);
+  const { reactFlowInstance, setReactFlowInstance } = useContext(rfContext);
+
+  // Fetching workspace state to populate the initial nodes/edges.
+  // Where/how the low-level ReactFlow JSON will be persisted is TBD.
+  // TODO: update when that design is finalized
+  const storedComponents = useSelector(
+    (state: AppState) => state.workspace.components
+  );
+  const { rfNodes, rfEdges } = convertToReactFlowData(storedComponents);
+  const [nodes, setNodes, onNodesChange] = useNodesState(rfNodes);
+  const [edges, setEdges, onEdgesChange] = useEdgesState(rfEdges);
+
+  const onConnect = useCallback(
+    (params) => {
+      setEdges((eds) => addEdge(params, eds));
+    },
+    // TODO: add customized logic to prevent connections based on the node's
+    // allowed inputs. If allowed, update that node state as well with the added
+    // connection details.
+    [setEdges]
+  );
+
+  const onDragOver = useCallback((event) => {
+    event.preventDefault();
+    event.dataTransfer.dropEffect = 'move';
+  }, []);
+
+  const onDrop = useCallback(
+    (event) => {
+      event.preventDefault();
+      // Get the node info from the event metadata
+      const nodeData = event.dataTransfer.getData('application/reactflow');
+
+      // check if the dropped element is valid
+      if (typeof nodeData === 'undefined' || !nodeData) {
+        return;
+      }
+
+      // Fetch bounds based on the ref'd div component, adjust as needed.
+      // TODO: remove hardcoded bounds and fetch from a constant somewhere
+      // @ts-ignore
+      const reactFlowBounds = reactFlowWrapper.current.getBoundingClientRect();
+      // @ts-ignore
+      const position = reactFlowInstance.project({
+        x: event.clientX - reactFlowBounds.left - 80,
+        y: event.clientY - reactFlowBounds.top - 90,
+      });
+
+      // TODO: remove hardcoded values when more component info is passed in the event.
+      // Only keep the calculated 'positioning' field.
+      const newNode = {
+        // TODO: generate ID based on the node data maybe
+        id: Date.now().toFixed(),
+        type: nodeData.type,
+        position,
+        data: { label: nodeData.label },
+        style: {
+          background: 'white',
+        },
+      };
+
+      // TODO: add logic to add node into the redux datastore
+
+      setNodes((nds) => nds.concat(newNode));
+    },
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [reactFlowInstance]
+  );
+
+  // Initialization hook
+  useEffect(() => {
+    // TODO: fetch the nodes/edges dynamically (loading existing flow,
```

Comment:
Will all these options live here? So basically we will have some check if we have some current flowID in our state? or from our in house fixed templates and etc.

Reply:
We will need to have some initialization logic to determine whether the page is loaded from an existing workflow, a new workflow, or (potentially) redirected from a sample template as a starting point, similar to Flowise.

For loading an existing workflow, we can parse the url for a workflow ID, fetch workflow details that contain the set of nodes/edges in the UI metadata, and load them into the reactflow instance via setNodes() / setEdges().

For starting with a blank canvas, we can simply set nodes/edges to empty arrays.

For redirecting from a sample template, we can focus on that as the UX design becomes more clear on that flow and if it is something we plan to do.